server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryTests.java

```diff
@@ -445,6 +450,37 @@ public void testGetRepositoryDataThreadContext() {
         future.actionGet(10, TimeUnit.SECONDS);
     }
 
+    public void testGetRepositoryDataForking() {
+        final var forkedListeners = new ArrayList<Runnable>();
+        final var future = new PlainActionFuture<Void>();
+        try (var listeners = new RefCountingListener(future)) {
+            final var repo = setupRepo();
+            final int threads = between(1, 5);
+            final var barrier = new CyclicBarrier(threads);
+            final var threadPool = client().threadPool();
+            final var testThread = Thread.currentThread();
+            final var resultsCountDown = new CountDownLatch(threads);
+            for (int i = 0; i < threads; i++) {
+                threadPool.generic().execute(ActionRunnable.wrap(listeners.acquire(), l -> {
+                    final var callingThread = Thread.currentThread();
+                    safeAwait(barrier);
+                    repo.getRepositoryData(runnable -> {
+                        forkedListeners.add(runnable);
+                        resultsCountDown.countDown();
+                    }, l.map(repositoryData -> {
+                        assertThat(Thread.currentThread(), oneOf(callingThread, testThread));
+                        resultsCountDown.countDown();
```

Comment:
Nit: I wonder whether we can make this assertion stricter. If this is directly invoked without going through the stub executor, the thread is going to be `callingThread`. When `forkedListeners.forEach` is invoked later, the thread is `testThread`. Also, the `resultsCountDown` keeps counting down after it reaches 0 when `forkedListeners.forEach` is called. It does not affect the test. But feels a bit weird?

Reply:
Yeah that behaviour was deliberate but we can make this a bit more explicit indeed, see bdc9be7bd031e6357353d1b3dbeaac80fd776e82.